Change fairy soul beacon colors based on distance, always track found souls

CraftyOldMiner opened the pull request:
Play testing (and code review) feedback on this change is appreciated.

- Fairy soul beacons now change color based on their distance
- Separated logic for tracking found souls from showing beacons. Tracking is turned on by default.
- Stop recalculating foundSouls each time tick() is called
- Refactored most of the FairySouls class to eliminate duplicate code and improve readability. Profile loading saving logic was retained.
- Change the FairySouls class to use a static instance instead of being a static class
- Remove redundant Forge EVENT_BUS register calls

NopoTheGamer reviewed (commented):
- If you do `/neusouls on` while "Track Fairy Souls" is off there should be an error
- "Show waypoints" seems to have no effect
- The colors do change when I move around

CraftyOldMiner commented:
> * If you do `/neusouls on` while "Track Fairy Souls" is off there should be an error
> * "Show waypoints" seems to have no effect
> * The colors do change when I move around

I fixed the "Show Waypoints" toggle and make it so turning tracking on/off doesn't require a lobby change. I also added on error like you suggested if you try and use /neusouls when tracking is off.

romangraef commented on src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/FairySouls.java:
```diff
@@ -21,44 +21,273 @@
+			missingSoulsDistanceSqMap.put(distSq, pos);
+		}
+		closestMissingSouls.clear();
+		if (missingSoulsDistanceSqMap.isEmpty()) {
```
Note: This return seems pointless.

CraftyOldMiner replied:
The return is to make it clear that there's no more work to be done if there are no missing souls instead of having to continue to read the rest of the method to fully understand what happens in this case.

romangraef commented on src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/FairySouls.java:
```diff
@@ -21,44 +21,273 @@
+	private HashMap<String, HashMap<String, Set<Integer>>> allProfilesFoundSouls = new HashMap<>();
+	private List<BlockPos> allSoulsInCurrentLocation;
+	private Set<Integer> foundSoulsInLocation;
+	private TreeMap<Double, BlockPos> missingSoulsDistanceSqMap;
```
I don't really like the idea of:

 - using `double` as a `Map` key.
 - storing the distance of missing souls in a map with the distance as a key (possible loss of souls because of duplicate keys is very unlikely, but this still *feels* wrong)
 - using the implicit sorting of `TreeMap` to sort a collection.

None of those are a deal breaker, but since we are using this only in one method (minus initialization), i would at least inline this variable to contain whatever this is.

CraftyOldMiner replied:
This PR tries to strike the right balance between cleaning up the code, adding the desired features, and minimizing the chance of regression. While I agree with your comment, I chose not to change this part of the design as part of this PR. I just renamed / moved the existing logic; see line 148 in the original file where distanceSqMap was defined.

romangraef commented on src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/FairySouls.java:
```diff
@@ -21,44 +21,273 @@
+		}
+	}
+
+	private int interpolateColors(int color1, int color2, double factor) {
+		int r1 = ((color1 >> 16) & 0xff);
+		int g1 = ((color1 >> 8) & 0xff);
+		int b1 = (color1 & 0xff);
+
+		int r2 = (color2 >> 16) & 0xff;
+		int g2 = (color2 >> 8) & 0xff;
+		int b2 = color2 & 0xff;
+
+		int r3 = r1 + (int)Math.round(factor * (r2-r1));
+		int g3 = g1 + (int)Math.round(factor * (g2-g1));
+		int b3 = b1 + (int)Math.round(factor * (b2-b1));
+
+		return (r3 & 0xff) << 16 |
+			(g3 & 0xff) << 8 |
+			(b3 & 0xff);	}
+
+	private double normalize(double value, double min, double max) {
+		return ((value - min) / (max - min));
+	}
```
Those two methods feel like they could belong in a utility class. (on the other hand our `Utils` class is extremely bloated and should probably be spread out over multiple files)

CraftyOldMiner replied:
Yeah, I actually looked for a good utility class to put these in and wasn't up for doing the cleanup so I left them here rather than adding them to the mess.

romangraef commented on src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/FairySouls.java:
```diff
@@ -21,44 +21,273 @@
+
+					locationSouls.add(new BlockPos(x, y, z));
+				}
+			} catch (Exception ignored) {
```
I think ignoring this exception is going to make debugging harder if a user ends up with a corrupted fairy soul list. printStackTrace should be fine.

CraftyOldMiner replied:
Agreed. This is part of the old code that I didn't spend the time to clean up. That said, I added the printStackTrace here and on line 330 as well so there's zero risk of regression.

romangraef commented on src/main/java/io/github/moulberry/notenoughupdates/options/NEUConfig.java:
```diff
@@ -129,6 +132,10 @@ public void executeRunnable(int runnableId) {
 				return;
+			case 15:
+				String command = NotEnoughUpdates.INSTANCE.config.misc.fariySoul ? "/neusouls on" : "/neusouls off";
+				ClientCommandHandler.instance.executeCommand(Minecraft.getMinecraft().thePlayer, command);
```
Confused: Is there a reason to use a command for this, but not for `setTrackFairySouls`? this seems fine, still.

CraftyOldMiner replied:
I did that for two reasons:
1. It provides consistent behavior between the pre-existing command and using the menu system to toggle the option.
2. The command can fail and I didn't see an obvious, lightweight way to show an error in the menu system and there was no reason to duplicate the error handling code.
